sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp

```diff
@@ -0,0 +1,212 @@
+
+    if (is_consumer)
+      __ESIMD_ENS::named_barrier_wait(b);
+
+    auto val = experimental_esimd::lsc_slm_block_load<int, VL>(
+        slm_base + off * sizeof(int));
+    // and storing it to output surface
+    experimental_esimd::lsc_fence();
+    experimental_esimd::lsc_block_store<int, VL>(o + off + j * slm_size, val);
+    experimental_esimd::lsc_fence();
+  }
+}
+
+[[intel::device_indirectly_callable]] SYCL_EXTERNAL void __regcall SIMD_CALLEE_nbarrier(
+#ifdef USE_ACC_PTR
+    local_accessor<int, 1> *LocalAcc,
+#else
+    local_accessor<int, 1> LocalAcc,
+#endif
+    int localID, int *o) SYCL_ESIMD_FUNCTION {
+#ifdef USE_ACC_PTR
+  ESIMD_CALLEE_nbarrier(*LocalAcc, localID, o);
+#else
+  ESIMD_CALLEE_nbarrier(LocalAcc, localID, o);
+#endif
+}
+
+int main() {
+  auto q = queue{gpu_selector_v};
+  auto dev = q.get_device();
+  std::cout << "Running on " << dev.get_info<sycl::info::device::name>()
+            << "\n";
+
+  auto deviceSLMSize =
+      dev.template get_info<sycl::info::device::local_mem_size>();
+  std::cout << "Local Memory Size: " << deviceSLMSize << std::endl;
+
+  // The test is going to use Size elements of int type.
+  if (deviceSLMSize < Size * sizeof(int)) {
+    // Report an error - the test needs a fix.
+    std::cerr << "Error: Test needs more SLM memory than device has"
+              << std::endl;
+    return -1;
+  }
+
+  auto *out = malloc_shared<int>(Size, q);
+  for (int i = 0; i < Size; i++) {
+    out[i] = -1;
+  }
+
+  try {
+    // workgroups
+    sycl::range<1> GlobalRange{Size};
+    // threads in each group
+    sycl::range<1> LocalRange{Size / Groups};
+    sycl::nd_range<1> Range{GlobalRange * LocalRange, LocalRange};
+
+    auto e = q.submit([&](handler &cgh) {
+      auto LocalAcc = local_accessor<int, 1>(Size, cgh);
+      cgh.parallel_for<KernelID>(
+          nd_range<1>(GlobalRange, LocalRange),
+          // This test requires an explicit specification of the subgroup size
+          [=](nd_item<1> item) [[intel::reqd_sub_group_size(VL)]] {
+            sycl::group<1> g = item.get_group();
+            sycl::sub_group sg = item.get_sub_group();
+
+            // Thread local ID in ESIMD context
+            int localID = sg.get_group_linear_id();
+
+            // SLM init
+            uint32_t slmID = item.get_local_id(0);
+            auto LocalAccCopy = LocalAcc;
+            LocalAccCopy[slmID] = -1;
+            item.barrier();
+
+#ifdef USE_ACC_PTR
+            auto LocalAccArg = uniform{&LocalAccCopy};
+#else
+            auto LocalAccArg = uniform{LocalAccCopy};
+#endif
+
+            invoke_simd(sg, SIMD_CALLEE_nbarrier, LocalAccArg, uniform{localID},
+                        uniform{out});
+          });
+    });
+    e.wait();
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    free(out, q);
+    return -1;
+  }
+
+  bool passed = true;
+  for (int i = 0; i < Size; i++) {
+    int etalon = 2;
+    if (i < Size / 4)
```

Comment:
can we make the computation of etalon clearer? i don't understand this, i just see random calculations

Reply:
Also reworked and with some comments